### ISSUE TITLE
Looking great!

### ISSUE THREAD

**Issue by stuzoo** [gui-dos/DiaBLE#1]
Hi Guido, 

DiaBLE looks great, I have been biohacking and playing with the libre freestyle to look at the data. I had some questions for you, can I ask them here? 

Thanks, 
Stuart

**Comment by gui-dos** [comment]
Yes, I am especially interested in fresh new ideas not discussed elsewhere.

**Comment by stuzoo** [comment]
I'm really interested in seeing what kind of raw data you can get out of the libre sensor, previously I was using a script to go download my values directly from libreview into my application which was a bit messy. 

I am about to start testing with a new sensor, it looks like if I should be able to test with DiaBLE and also use the libre app at the same time. Then I can do some comparisons. Have you done any comparisons of the readings?

**Comment by gui-dos** [comment]
That’s exactly the same reason why I started to play with SwiftUI on my iPad a year ago.

Bubble’s team assured me that their glucose.space server wouldn’t be taken down for open-source projects therefore I could display two curves, the raw glucose values and the calibrated ones provided by the LibreLink libraries which they are making available online to iPhone users.

It was then @Flameeyes that suggested me to have a look at the last issue of PoC||GTFO which details the exploit you can leverage on to dump all the memory of the Libre 1 microcontroller. Diego has been working for years to reverse engineer the USB FreeStyle protocols but the Libre2 one is still to be understood.

**Comment by stuzoo** [comment]
So much info out there from @Flameeyes <https://github.com/Flameeyes> amazing. 

Did you see this one nutrisense.io <http://nutrisense.io/>, from the youtube videos they are using the the Libre freestyle and they are scanning with their own app, I wonder what algorithm they are using? , I suppose they are just giving nutrition advice so they might not need to be very accurate? 

On 8 Sep 2020, at 12:08, Guido Soranzio <notifications@github.com> wrote:


That’s exactly the same reason why I started to play with SwiftUI on my iPad a year ago.

Bubble’s team assured me that their glucose.space server wouldn’t be taken down for open-source projects therefore I could display two curves, the raw glucose values and the calibrated ones provided by the LibreLink libraries which they are making available online to iPhone users.

It was then @Flameeyes <https://github.com/Flameeyes> that suggested me to have a look at the last issue of PoC||GTFO which details the exploit you can leverage on to dump all the memory of the Libre 1 microcontroller. Diego has been working for years to reverse engineer the USB FreeStyle protocols but the Libre2 one is still to be understood.

**Comment by stuzoo** [comment]
Hi Guido, 

Just wondering if you were having any issues with the calibrateSensor endpoint? 

{
    "errcode": -1,
    "msg": "java.lang.NullPointerException\n\tat com.demo.index.Utils.getSlope(Utils.java:164)\n\tat com.demo.index.IndexController.calibrateSocketResult(IndexController.java:792)\n\tat com.demo.index.IndexController.calibrateSensor(IndexController.java:715)\n\tat sun.reflect.GeneratedMethodAccessor8.invoke(Unknown Source)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat com.jfinal.aop.Invocation.invoke(Invocation.java:97)\n\tat com.jfinal.core.ActionHandler.handle(ActionHandler.java:90)\n\tat com.jfinal.core.JFinalFilter.doFilter(JFinalFilter.java:89)\n\tat io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)\n\tat io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)\n\tat io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)\n\tat io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)\n\tat io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)\n\tat io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)\n\tat io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132)\n\tat io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)\n\tat io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)\n\tat io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)\n\tat io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)\n\tat io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)\n\tat io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)\n\tat io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)\n\tat io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)\n\tat io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)\n\tat io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)\n\tat io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)\n\tat io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)\n\tat io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)\n\tat io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)\n\tat io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)\n\tat io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)\n\tat io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)\n\tat io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)\n\tat io.undertow.server.Connectors.executeRootHandler(Connectors.java:364)\n\tat io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:748)\n"
}

On 8 Sep 2020, at 13:01, stuart brock <brocksd@gmail.com> wrote:

So much info out there from @Flameeyes <https://github.com/Flameeyes> amazing. 

Did you see this one nutrisense.io <http://nutrisense.io/>, from the youtube videos they are using the the Libre freestyle and they are scanning with their own app, I wonder what algorithm they are using? , I suppose they are just giving nutrition advice so they might not need to be very accurate? 

On 8 Sep 2020, at 12:08, Guido Soranzio <notifications@github.com <mailto:notifications@github.com>> wrote:


That’s exactly the same reason why I started to play with SwiftUI on my iPad a year ago.

Bubble’s team assured me that their glucose.space server wouldn’t be taken down for open-source projects therefore I could display two curves, the raw glucose values and the calibrated ones provided by the LibreLink libraries which they are making available online to iPhone users.

It was then @Flameeyes <https://github.com/Flameeyes> that suggested me to have a look at the last issue of PoC||GTFO which details the exploit you can leverage on to dump all the memory of the Libre 1 microcontroller. Diego has been working for years to reverse engineer the USB FreeStyle protocols but the Libre2 one is still to be understood.

**Comment by gui-dos** [comment]
Yes, it often happens. That's the second reason why I wrote DiaBLE: display the full error messages instead of freezing or crashing  like the other apps... :P
